src/transactions/TransactionFrame.cpp

```diff
@@ -1544,19 +1386,20 @@ TransactionFrame::removeAccountSigner(AbstractLedgerTxn& ltxOuter,
     }
 }
 
-bool
+std::pair<bool, MutableTxResultPtr>
 TransactionFrame::checkValidWithOptionallyChargedFee(
     Application& app, AbstractLedgerTxn& ltxOuter, SequenceNumber current,
     bool chargeFee, uint64_t lowerBoundCloseTimeOffset,
-    uint64_t upperBoundCloseTimeOffset)
+    uint64_t upperBoundCloseTimeOffset) const
 {
     ZoneScoped;
-    mCachedAccount.reset();
+    mCachedAccountPreProtocol8.reset();
 
     if (!XDRProvidesValidFee())
     {
-        getResult().result.code(txMALFORMED);
-        return false;
+        auto txResult = createTxResult();
+        txResult->setInnermostResultCode(txMALFORMED);
```

Comment:
This is necessary to get rid of the dynamic casts we had previously that downcast `MutableTransactionResultBase` to `FeeBumpTransactionResult`. Currently, `FeeBumpTransactionFrame` calls the inner `TransactionFrame` functions with the expectation that the `TransactionFrame` will only update the inner result. Fee Bump expects every call to `TransactionFrame` to only modify the inner result, so I think it makes the most sense for all `TransactionFrame` functions to only ever modify the inner most result for consistency.

Reply:
>  Currently, FeeBumpTransactionFrame calls the inner TransactionFrame functions with the expectation that the TransactionFrame will only update the inner result.

But TxFrame doesn't ever deal with the fee bump result, does it? At least that was my understanding when reading the code.

Reply:
It does. Whenever `FeeBumpTransactionFrame` calls an innerTx frame function, it passes a FeeBumpTransactionResult to the innerTx (such as [this](https://github.com/stellar/stellar-core/blob/a8d87311d720f0efded62d6eb1e6473c9a30e5c9/src/transactions/FeeBumpTransactionFrame.cpp#L107)) . Previously, I used a dynamic cast here to only pass the inner `MutableTransactionResult` to `TransactionFrame`. After refactoring out the cast, I now pass the full `FeeBumpMutableResult` with the expectation that `TransactionFrame` will access the inner result via `innerMostResult`

Reply:
Interesting, I've looked at `checkValid` which does this differently and has no way of using the wrong tx result type. Is there a particular reason for `apply` to not work like `checkValidWithOptionallyChargedFee` and not return the result instead of modifying the mutable input result?

Reply:
Yeah, so there's essentially two "entry points" into the `MutableTransactionResult` flow. The first is via `checkValid`. This is only called during TxQueue submission and TxSet formation, and the returned TxResult is thrown away after forming the TxSet.

The 2nd entry point is `processFeeSeqNum`. This flow is in the actual application flow, where `LedgerManager` first calls `processFeeSeqNum` to generate the TxResult, then gives that result to `apply`, then to `processPostApply`. The issue is, `processFeeSeqNum` makes changes to txResult (namely sets the feeCharged) that need to persist during the following calls to `apply` and `processPostApply`. It looks like the `feeCharged` value is the only state in TxResult that `processFeeSeqNum` and `processPostApply` modify so in theory we could construct a new TxResult in apply and just copy the old value over, but that seems pretty footgunny as logically all these functions should mutate the same TxResult. Personally I'd like to keep the interface as is, where the two "entry points" return a new result, where the non entry points mutate the generated TxResult.

Reply:
I feel like this just points at some issues with abstractions we have. Maybe fee charged shouldn't be (ab-)used as a part of the mutable result (it also has pretty messy semantics in tx submission flow). Basically `processFeeSeqNum` should probably only return the fee charged, and `apply` should return the result of applying the transaction, and then we should tie these together in XDR. But I'm fine with leaving this as is for this PR; I'll maybe do one more refactoring iteration after it's merged and try to clean this up more.